qingstor/qsctl/commands/transfer.py

```diff
@@ -230,21 +242,41 @@ def download_file(cls, options):
     def write_local_file(cls, local_path, bucket, key, options):
         cls.validate_bucket(bucket)
         current_bucket = cls.client.Bucket(bucket, cls.bucket_map[bucket])
-        resp = current_bucket.get_object(key)
+        record = cls.recorder.get_record(bucket, key, RECORD_DOWNLOAD)
+        completed = 0
+        if record != "":
+            try:
+                completed = int(record)
+            except ValueError:
+                print ("Error: Invalid completed size record %s" % record)
+                sys.exit(-1)
+        if completed > 0:
+            resp = current_bucket.get_object(key, range="bytes=%d-" % completed)
+            uni_print("Resume downloading key <%s>" % key)
+        else:
+            resp = current_bucket.get_object(key)
         if resp.status_code in (HTTP_OK, HTTP_OK_PARTIAL_CONTENT):
             cls.validate_local_path(local_path)
+            write_path = local_path
+            open_flag = "wb"
             if key[-1] != "/":
                 content_length = resp.headers["Content-Length"]
```

Comment:
https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html   "The Content-Length entity-header field indicates the size of the entity-body"